### ISSUE TITLE
poor results on the TUM dataset 。

### ISSUE THREAD

**Issue by mightyhao** [haleqiu/AirIMU#11]
Thanks for your great work!I am currently working with the AirIMU model and have trained it using the TUM dataset. However, I am experiencing poor results, particularly with the inte_error_compare and state_error_compare outputs. I have attached the model trained on the TUM dataset, trained result as follows:
train loss: -0.186567 test loss: -0.185325
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 742/742 [05:40<00:00,  2.18it/s]
evaluating: position loss 1.148859, rotation loss 0.016350, vel losses 0.371750
eval pos: 1.148859 eval rot: 0.016350
training epoch: 048, losses: -0.193718, position, 0.151465 rotation 0.006946, pred_rot 0.000019, pred_cov0.007398: 100%|███████████████████| 36/36 [02:30<00:00,  4.17s/it]
testing losses: -0.182188, position, 0.168962 rotation 0.006489, vel 0.061111: 100%|█████████████████████████████████████████████████████████| 4/4 [00:03<00:00,  1.19it/s]
train loss: -0.187098 test loss: -0.182188
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 742/742 [05:06<00:00,  2.42it/s]
evaluating: position loss 1.187331, rotation loss 0.016090, vel losses 0.384744
eval pos: 1.187331 eval rot: 0.016090
training epoch: 049, losses: -0.194848, position, 0.149035 rotation 0.006859, pred_rot 0.000017, pred_cov0.005162: 100%|███████████████████| 36/36 [02:29<00:00,  4.16s/it]
testing losses: -0.182290, position, 0.171431 rotation 0.006512, vel 0.061502: 100%|█████████████████████████████████████████████████████████| 4/4 [00:03<00:00,  1.18it/s]
train loss: -0.187602 test loss: -0.182290
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 742/742 [05:12<00:00,  2.37it/s]
evaluating: position loss 1.163970, rotation loss 0.016012, vel losses 0.378961
eval pos: 1.163970 eval rot: 0.016012
wandb:                                                                                

wandb: 
wandb: Run summary:
wandb:      eval/pos_dist 1.16397
wandb:      eval/pos_loss 1.16397
wandb:      eval/rot_dist 0.01601
wandb:      eval/rot_loss 0.01601
wandb:      eval/vel_dist 0.37896
wandb:      eval/vel_loss 0.37896
wandb:                 lr 0.001
wandb:          test/loss -0.18229
wandb:      test/pos_loss 0.17143
wandb:  test/pred_cov_pos 0.00682
wandb:  test/pred_cov_rot 2e-05
wandb:  test/pred_cov_vel 0.00103
wandb:      test/rot_loss 0.00651
wandb:      test/vel_loss 0.0615
wandb:         train/loss -0.1876
wandb:     train/pos_loss 0.14903
wandb: train/pred_cov_pos 0.0064
wandb: train/pred_cov_rot 2e-05
wandb: train/pred_cov_vel 0.00097
wandb:     train/rot_loss 0.00686
wandb:     train/vel_loss 0.05252

and the results are shown in the attached JSON file is as follows:

[
    {
        "name": "dataset-room2_512_16",
        "use_gt_rot": false,
        "AOE(raw)": 2.48331096136933,
        "ATE(raw)": 309.9386655143033,
        "AVE(raw)": 6.915526182640868,
        "ROE(raw)": 1.0084754696578992,
        "RTE(raw)": 0.10585664625035328,
        "RVE(raw)": 0.1756534599813752,
        "RP_RMSE(raw)": 0.190841797011662,
        "RV_RMSE(raw)": 0.32052022733649216,
        "RO_RMSE(raw)": 1.8824918754341047,
        "O_RMSE(raw)": 3.024278446390196,
        "AOE(AirIMU)": 1.5821726831792424,
        "ATE(AirIMU)": 272.2560563378184,
        "AVE(AirIMU)": 6.722642719819769,
        "ROE(AirIMU)": 0.9900209244425077,
        "RTE(AirIMU)": 0.10286696012252976,
        "RVE(AirIMU)": 0.1699432410011462,
        "RP_RMSE(AirIMU)": 0.1906377701295835,
        "RV_RMSE(AirIMU)": 0.32052880684088086,
        "RO_RMSE(AirIMU)": 1.8763327701796197,
        "O_RMSE(AirIMU)": 1.9530199145993792
    },
    {
        "name": "dataset-room4_512_16",
        "use_gt_rot": false,
        "AOE(raw)": 1.540609030342215,
        "ATE(raw)": 193.1219888201075,
        "AVE(raw)": 5.604680926999612,
        "ROE(raw)": 0.4963133284732377,
        "RTE(raw)": 0.06302456031956952,
        "RVE(raw)": 0.11457781500764185,
        "RP_RMSE(raw)": 0.10931900897080232,
        "RV_RMSE(raw)": 0.17423298053996417,
        "RO_RMSE(raw)": 0.8612122200354659,
        "O_RMSE(raw)": 1.6912857275186046,
        "AOE(AirIMU)": 1.5182884081920405,
        "ATE(AirIMU)": 255.1958328714938,
        "AVE(AirIMU)": 8.96275326700491,
        "ROE(AirIMU)": 0.4852878804533531,
        "RTE(AirIMU)": 0.058287033813320135,
        "RVE(AirIMU)": 0.10293506342070897,
        "RP_RMSE(AirIMU)": 0.10796548194223628,
        "RV_RMSE(AirIMU)": 0.1657293040803919,
        "RO_RMSE(AirIMU)": 0.8427513426448806,
        "O_RMSE(AirIMU)": 1.7262002040364062
    },
    {
        "name": "dataset-room6_512_16",
        "use_gt_rot": false,
        "AOE(raw)": 1.1108805245351439,
        "ATE(raw)": 356.32847673261176,
        "AVE(raw)": 8.871778452179399,
        "ROE(raw)": 0.5495216317023245,
        "RTE(raw)": 0.4045390675602481,
        "RVE(raw)": 0.5925928275208284,
        "RP_RMSE(raw)": 0.47848576650538055,
        "RV_RMSE(raw)": 0.7006707895438172,
        "RO_RMSE(raw)": 1.4809729820027242,
        "O_RMSE(raw)": 1.814046805826966,
        "AOE(AirIMU)": 2.157903870828523,
        "ATE(AirIMU)": 328.4892069205532,
        "AVE(AirIMU)": 7.767304617256555,
        "ROE(AirIMU)": 0.5354635364186967,
        "RTE(AirIMU)": 0.4014274082994396,
        "RVE(AirIMU)": 0.5913399964983228,
        "RP_RMSE(AirIMU)": 0.4749228194337617,
        "RV_RMSE(AirIMU)": 0.7004630966168738,
        "RO_RMSE(AirIMU)": 1.4740604854016583,
        "O_RMSE(AirIMU)": 2.878123551028877
    },
    {
        "name": "dataset-room1_512_16",
        "use_gt_rot": false,
        "AOE(raw)": 1.4187193092392236,
        "ATE(raw)": 224.05333861074493,
        "AVE(raw)": 4.8967928677999675,
        "ROE(raw)": 0.6946434004761611,
        "RTE(raw)": 0.06934604803466082,
        "RVE(raw)": 0.12036826157142064,
        "RP_RMSE(raw)": 0.16461842006644273,
        "RV_RMSE(raw)": 0.26000240591739854,
        "RO_RMSE(raw)": 1.6145092222059059,
        "O_RMSE(raw)": 2.3820463023956644,
        "AOE(AirIMU)": 2.0559409107953903,
        "ATE(AirIMU)": 70.93059649306558,
        "AVE(AirIMU)": 3.1130422788606404,
        "ROE(AirIMU)": 0.6311477934512721,
        "RTE(AirIMU)": 0.06061297729038395,
        "RVE(AirIMU)": 0.10068604708124924,
        "RP_RMSE(AirIMU)": 0.16169444294295274,
        "RV_RMSE(AirIMU)": 0.2515812199609861,
        "RO_RMSE(AirIMU)": 1.6138927240203513,
        "O_RMSE(AirIMU)": 2.642126250187713
    },
    {
        "name": "dataset-room3_512_16",
        "use_gt_rot": false,
        "AOE(raw)": 1.796761575828253,
        "ATE(raw)": 552.991748556449,
        "AVE(raw)": 12.41507565304104,
        "ROE(raw)": 0.7571651762279796,
        "RTE(raw)": 0.08670657163312742,
        "RVE(raw)": 0.15630715925467356,
        "RP_RMSE(raw)": 0.16516990036299709,
        "RV_RMSE(raw)": 0.27702045832218813,
        "RO_RMSE(raw)": 1.1352422373000228,
        "O_RMSE(raw)": 2.0948732000994545,
        "AOE(AirIMU)": 2.358007695679839,
        "ATE(AirIMU)": 406.91120218111655,
        "AVE(AirIMU)": 9.731401342157769,
        "ROE(AirIMU)": 0.7081502134444494,
        "RTE(AirIMU)": 0.08266566230622271,
        "RVE(AirIMU)": 0.1450900620951414,
        "RP_RMSE(AirIMU)": 0.16295711897909457,
        "RV_RMSE(AirIMU)": 0.27347318862389963,
        "RO_RMSE(AirIMU)": 1.1006456035686472,
        "O_RMSE(AirIMU)": 2.879972556835092
    },
    {
        "name": "dataset-room5_512_16",
        "use_gt_rot": false,
        "AOE(raw)": 0.6607448585619204,
        "ATE(raw)": 187.23084888230898,
        "AVE(raw)": 4.151465227997613,
        "ROE(raw)": 0.5052652634177358,
        "RTE(raw)": 0.05139710950178838,
        "RVE(raw)": 0.0879977217825901,
        "RP_RMSE(raw)": 0.06441897658088408,
        "RV_RMSE(raw)": 0.10180563819443385,
        "RO_RMSE(raw)": 0.6044280678923449,
        "O_RMSE(raw)": 0.7637142274110883,
        "AOE(AirIMU)": 1.6124729226650798,
        "ATE(AirIMU)": 432.4163120116147,
        "AVE(AirIMU)": 11.874582829780003,
        "ROE(AirIMU)": 0.4023251842851207,
        "RTE(AirIMU)": 0.04230045881954144,
        "RVE(AirIMU)": 0.07262399779792439,
        "RP_RMSE(AirIMU)": 0.05682918596019412,
        "RV_RMSE(AirIMU)": 0.08952093609631293,
        "RO_RMSE(AirIMU)": 0.5150862261300284,
        "O_RMSE(AirIMU)": 1.7752077098609076
    },
    {
        "name": "dataset-room1_512_16",
        "use_gt_rot": false,
        "AOE(raw)": 1.4187193092392236,
        "ATE(raw)": 224.05333861074493,
        "AVE(raw)": 4.8967928677999675,
        "ROE(raw)": 0.6946434004761611,
        "RTE(raw)": 0.06934604803466082,
        "RVE(raw)": 0.12036826157142064,
        "RP_RMSE(raw)": 0.16461842006644273,
        "RV_RMSE(raw)": 0.26000240591739854,
        "RO_RMSE(raw)": 1.6145092222059059,
        "O_RMSE(raw)": 2.3820463023956644,
        "AOE(AirIMU)": 2.0559409107953903,
        "ATE(AirIMU)": 70.93059649306558,
        "AVE(AirIMU)": 3.1130422788606404,
        "ROE(AirIMU)": 0.6311477934512721,
        "RTE(AirIMU)": 0.06061297729038395,
        "RVE(AirIMU)": 0.10068604708124924,
        "RP_RMSE(AirIMU)": 0.16169444294295274,
        "RV_RMSE(AirIMU)": 0.2515812199609861,
        "RO_RMSE(AirIMU)": 1.6138927240203513,
        "O_RMSE(AirIMU)": 2.642126250187713
    }
]

Could you please help me understand the possible reasons for these poor results? Additionally, I would appreciate your guidance on how to adjust the training parameters for the TUM dataset to achieve more reasonable and improved outcomes.

Looking forward to your advice!

Best regards,

**Comment by mightyhao** [comment]
![Uploading dataset-room1_512_16_orientation_compare.png…]()

**Comment by mightyhao** [comment]
![dataset-room1_512_16_orientation_compare](https://github.com/user-attachments/assets/5cefd258-d841-4365-ab5c-41e076dcd26d)
![Uploading dataset-room1_512_16inte_error_compare.png…]()
![Uploading dataset-room1_512_16state_error_compare.png…]()

**Comment by mightyhao** [comment]
dataset-room1_512_16  eval results:
![Uploading dataset-room1_512_16_orientation_compare.png…]()
![Uploading dataset-room1_512_16inte_error_compare.png…]()
![Uploading dataset-room1_512_16state_error_compare.png…]()

**Comment by mightyhao** [comment]
dataset-room1_512_16 eval results:
![dataset-room1_512_16_orientation_compare](https://github.com/user-attachments/assets/eef41141-4c3b-4ffb-9bf2-eb3f0a05c842)
![dataset-room1_512_16inte_error_compare](https://github.com/user-attachments/assets/82304118-5eb9-4ddf-b90f-5f5c2ea8adbc)
![dataset-room1_512_16state_error_compare](https://github.com/user-attachments/assets/8755b653-6bac-4c52-98a3-31119e552333)

**Comment by haleqiu** [comment]
@mightyhao 
Hi, from your figure, both the ROE and RTE are improved. It would be great if you could detail which part you're not clear about.

The TUM dataset is a little bit small and difficult to train.

**Comment by haleqiu** [comment]
For example, in room2

"ROE(raw)": 1.0084754696578992,
"RTE(raw)": 0.10585664625035328,

"ROE(AirIMU)": 0.9900209244425077,
"RTE(AirIMU)": 0.10286696012252976,

This improvement is small because the 1s-movement is pretty tiny. If you integrate it for a long time, the error will be significant, for example, ATE.

**Comment by mightyhao** [comment]
> For example, in room2
> 
> "ROE(raw)": 1.0084754696578992, "RTE(raw)": 0.10585664625035328,
> 
> "ROE(AirIMU)": 0.9900209244425077, "RTE(AirIMU)": 0.10286696012252976,
> 
> This improvement is small because the 1s-movement is pretty tiny. If you integrate it for a long time, the error will be significant, for example, ATE.

Thank you very much for your reply. From the results, it seems that whether AirIMU is applied or not, the direction angle data obtained is similar to the original data. Logically, since TUMVI is a low-end IMU, wouldn't the angle error increase more after integrating the raw data? (For EuRoc data, the IMU is a higher-end model, and the raw data's ROE is quite poor. However, after processing with the model, the ROE is very good, and the position and velocity errors remain relatively small over time). I also analyzed the position and velocity results, and from the results, the position and velocity errors are relatively large. This indicates that AirIMU did not effectively reduce the impact of acceleration bias on the TUMVI, resulting in poor position and velocity outcomes after acceleration integration. Shouldn't AirIMU remove static biases and non-deterministic errors? Is it possible for the TUM training results to achieve similar performance to EuRoc? If so, could you provide some suggestions? Thank you!

**Comment by mightyhao** [comment]
And I directly used the IMU data for training and testing. The test results show that the ROE(AirIMU) for room2, room4, and room6 are 0.99°, 0.48°, and 0.53° respectively, with an average of 0.67°. However, the ROE(raw) in room2, room4, and room6 are 1.00°, 0.50°, and 0.55°, with an average of 0.68°. The paper mentions a value of 2.4936ROE (raw) ) on the TUMVI
dataset.

**Comment by haleqiu** [comment]
@mightyhao Again, like what I just mentioned. The size of the TUM dataset is pretty small, and the useful information is limited (the data with ground truth). We don't really think the model is trained sufficiently. Unfortunately, there is little thing we can do about this issue.

About the number in the paper: The TUM runs a Kalibr for their datasets. The IMU you get is something after Kalibr. If you want to access the raw data, please check their website and documentation.

**Comment by mightyhao** [comment]
> @mightyhao Again, like what I just mentioned. The size of the TUM dataset is pretty small, and the useful information is limited (the data with ground truth). We don't really think the model is trained sufficiently. Unfortunately, there is little thing we can do about this issue.
> 
> About the number in the paper: The TUM runs a Kalibr for their datasets. The IMU you get is something after Kalibr. If you want to access the raw data, please check their website and documentation.

Thank you for your reply, I understand now. I have another question: I have a Bosch BMI055 IMU, and I want to use data collected from this IMU to train the AirIMU model. Can the pre-trained model be applied to data from another IMU (also a BMI055) for denoising? Or do I need to collect as much data as possible from the same type of IMU to ensure the model generalizes well?

**Comment by haleqiu** [comment]
@mightyhao Actually, that's a good question. You may check my Discussion and future works section in our paper.

To be honest, I haven't try it. It is worth a try to explore some pre-train/fine-tune/online learning tricks. Please keep me update if you're interested in this problem.

Also, you should pay attention to the fact that sometimes the IMU contains deterministic errors. Sometimes the in-factory calibration covers that part. Sometimes there are not.

**Comment by mightyhao** [comment]
> @mightyhao Actually, that's a good question. You may check my Discussion and future works section in our paper.
> 
> To be honest, I haven't try it. It is worth a try to explore some pre-train/fine-tune/online learning tricks. Please keep me update if you're interested in this problem.
> 
> Also, you should pay attention to the fact that sometimes the IMU contains deterministic errors. Sometimes the in-factory calibration covers that part. Sometimes there are not.


Thank you, I look forward to your latest research and results
